### DISCUSSION TITLE
Core 2.4.0 - 2.5.0 // bin file larger (no code change)

### DISCUSSION THREAD

**Discussion by JeroenIoT** [stm32duino/Arduino_Core_STM32#2137] [General]
Hello,

I upgraded from core stm32duino core  **2.5.0** to **2.6.0** and my BIN file became larger 

* Went from approx. **46% to 50.x%**
* Working with STM32L082 in platformio environment

As I am using dual banks I need to reduce the code to below 50%.
And to be honest I have no idea where this increasement is coming from - any idea? 

Thanks a lot
deBuffel

**Comment by JeroenIoT** [comment]
Comparing the 2 versions by changing platform.io

```
[env]
; platform = ststm32 @ ~15.6.0        ; stm32duino core 2.4.0
platform = ststm32 @ ~16.1.0          ; stm32duino core 2.5.0

```
**16.1.0 (CORE 2.5.0)**
```
Checking size .pio\build\NCT_Latest\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [======    ]  55.9% (used 11448 bytes from 20480 bytes)
Flash: [=====     ]  50.6% (used 99436 bytes from 196608 bytes)
```

**15.6.0 (CORE 2.4.0)**
```
Checking size .pio\build\NCT_Latest\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [======    ]  55.7% (used 11408 bytes from 20480 bytes)
Flash: [=====     ]  46.7% (used 91816 bytes from 196608 bytes)
```

**Reply by fpistm** [reply]
Well BTW 2.4.0 and 2.5.0 L0 HAL have been updated and the gcc version. 
You should check symbol of each binary to see where it growth.
Or share them with debug symbol.

**Comment by JeroenIoT** [comment]
Thanks. I used https://www.sikorskiy.net/info/prj/amap/ to analyse the map files.
Most of the size of the modules became a bit smaller. 

There are 2 modules that became bigger (both of these modules in 2.5.0 arre roughly 20.000 (see below) compared to 15.000 in 2.4.0.

```
c:/users/jeroen/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp\libc_nano.a	20468	20523	280
c:/users/jeroen/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/10.3.1/thumb/v6-m/nofp\libgcc.a	20322	20322	195
```
So **libc_nano.a** and **libgcc.a** grew quite a bit. The question now is -  are there options to shrink them :-)
I use following build_flags that actially did shrink my bin files.
```
build_flags = 
                 -Wl,-Map,output.map
                -Os
                -flto -Os
                -specs=nano.specs -Os
```
However I still have a difference in size between core 2.4.0 and 2.5.0 (77072 vs 85616).
Do you know other alternatives?

**Reply by fpistm** [reply]
So it is link to the new toolchain. Unfortunately I don't know how to shrink more. Maybe avoid call of some functions linked to  libc

**Reply by JeroenIoT** [reply]
That is an ARM related lib afaik. I find it interesting to see that a new version consumes so much more.
I guess that code already to the limit will no longer fit in the STMs flash when they upgrade to the 2.5.0 core.
Will dig into it this week to see if I can find the reason ... post it here

**Reply by fpistm** [reply]
You should try with arduino ide we moved to gcc 12.

**Reply by fpistm** [reply]
Anyway I thought pio moved to 12 last week.

**Reply by fpistm** [reply]
I've just made some test to build blank sketch for Generic L082KZUX:

* Core 2.4.0 with arm-none-eabi-gcc 10.3.1-2.3: 
```
Sketch uses 9912 bytes (5%) of program storage space. Maximum is 196608 bytes.
Global variables use 888 bytes (4%) of dynamic memory, leaving 19592 bytes for local variables. Maximum is 20480 bytes.
```
* Core 2.5.0 with arm-none-eabi-gcc 10.3.1-2.3: 
```
Sketch uses 10040 bytes (5%) of program storage space. Maximum is 196608 bytes.
Global variables use 892 bytes (4%) of dynamic memory, leaving 19588 bytes for local variables. Maximum is 20480 bytes.
```
* Core 2.5.0 with arm-none-eabi-gcc 12.2.1-1.2: 
```
Sketch uses 10124 bytes (5%) of program storage space. Maximum is 196608 bytes.
Global variables use 1184 bytes (5%) of dynamic memory, leaving 19296 bytes for local variables. Maximum is 20480 bytes.
```

Difference is limited. In 2.5.0 we've updated the L0 Cube firmware it adds some limited size too.
So, I guess it is linked to something you used in your code bringing more code stuff from the C lib.
